src/librbd/operation/EnableFeaturesRequest.cc

```diff
@@ -159,7 +156,67 @@ Context *EnableFeaturesRequest<I>::handle_get_mirror_mode(int *result) {
     return handle_finish(*result);
   }
 
-  m_enable_mirroring = (mirror_mode == cls::rbd::MIRROR_MODE_POOL);
+  if (m_mirror_mode == cls::rbd::MIRROR_MODE_POOL) {
+    m_enable_mirroring = true;
+  }
+
+  send_get_mirror_image();
+  return nullptr;
+}
+
+template <typename I>
+void EnableFeaturesRequest<I>::send_get_mirror_image() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+
+  if (m_mirror_mode != cls::rbd::MIRROR_MODE_IMAGE) {
+    Context *ctx = create_context_callback<
+      EnableFeaturesRequest<I>,
+      &EnableFeaturesRequest<I>::handle_get_mirror_image>(this);
+    ctx->complete(-ENOENT);
+    return;
+  }
+
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_image_get_start(&op, image_ctx.id);
+
+  using klass = EnableFeaturesRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_callback<klass, &klass::handle_get_mirror_image>(this);
+  m_out_bl.clear();
+  int r = image_ctx.md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_out_bl);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *EnableFeaturesRequest<I>::handle_get_mirror_image(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": r=" << *result << dendl;
+
+  cls::rbd::MirrorImage mirror_image;
+  if (*result == 0) {
+    auto it = m_out_bl.cbegin();
+    *result = cls_client::mirror_image_get_finish(&it, &mirror_image);
+  } else if (*result == -ENOENT) {
+    *result = 0;
+  }
+
+  if (*result < 0) {
+    lderr(cct) << "failed to retrieve mirror image info: "
+               << cpp_strerror(*result) << dendl;
+    return handle_finish(*result);
+  }
+
+  if (mirror_image.mode == cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT) {
```

Comment:
Ah, I think I had this in my previous version but it seems I was confused with this your comment:

> If the mirror mode is journal and the pool mode is image, just prevent disabling the journal feature. Don't implicitly switch modes.

I missed you were talking about pool mode, and just did not want we automatically switch journal/snapshot mode when journaling feature enabled/disabled for a mirroring image from a pool in 'mirror image' mode.

So you think it is ok? Then I will revert back this.

Reply:
And I don't think we can't just enable journaling. As it was in my original code, we need to disable mirroring first and then re-enable it again so the rbd-mirror deleted the old mirroring image and created a new one. Or do you suppose rbd-mirror would handle this without recreating the non-primary image? Theoretically probably it could but do we want?

Reply:
We wouldn't be using this to switch mirroring modes, right? To switch between journal-based and image-based mirroring, you should have to disable mirroring and re-enable it on the image. When the mirror pool mode is in the legacy "pool" mode, we can assume we should default to journal-based mirroring.